RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc

```diff
@@ -0,0 +1,191 @@
+  float n5=0;
+  getLandauFrac(-tsShift5+25,-tsShift5+50,n5);
+
+  float ch3=corrCharge[3]/i3;
+  float ch4=(i3*corrCharge[4]-n3*corrCharge[3])/(i3*i4);
+  float ch5=(n3*n4*corrCharge[3]-i4*nn3*corrCharge[3]-i3*n4*corrCharge[4]+i3*i4*corrCharge[5])/(i3*i4*i5);
+
+  if (ch3<-3 && fNegStrat==HcalDeterministicFit::ReqPos) {
+    ch3=-3;
+    ch4=corrCharge[4]/i4;
+    ch5=(i4*corrCharge[5]-n4*corrCharge[4])/(i4*i5);
+  }
+  
+  if (ch5<-3 && fNegStrat==HcalDeterministicFit::ReqPos) {
+    ch4=ch4+(ch5+3);
+    ch5=-3;
+  }
+
+  if (fNegStrat==HcalDeterministicFit::FromGreg) {
+    if (ch3<-3) {
+      ch3=-3;
+      ch4=corrCharge[4]/i4;
+      ch5=(i4*corrCharge[5]-n4*corrCharge[4])/(i4*i5);
+    }
+    if (ch5<-3 && ch4>15) {
+      double ratio = (corrCharge[4]-ch3*i3)/(corrCharge[5]+3*i5);
+      if (ratio < 5 && ratio > 0.5) {
+	double invG = -13.11+11.29*TMath::Sqrt(2*TMath::Log(5.133/ratio));
+	float iG=0;
+	getLandauFrac(-invG,-invG+25,iG);
+	ch4=(corrCharge[4]-ch3*n3)/(iG);
+	ch5=-3;
+	tsShift4=invG;
+      }
+    }
+  }
+  
+  if (ch3<1) {// && (fNegStrat==HcalDeterministicFit::ReqPos || fNegStrat==HcalDeterministicFit::FromGreg)) {
+    ch3=0;
+  }
+  if (ch4<1) {// && (fNegStrat==HcalDeterministicFit::ReqPos || fNegStrat==HcalDeterministicFit::FromGreg)) {
+    ch4=0;
+  }
+  if (ch5<1) {// && (fNegStrat==HcalDeterministicFit::ReqPos || fNegStrat==HcalDeterministicFit::FromGreg)) {
+    ch5=0;
+  }
+  
+  // Print out 
+  HLTOutput.clear();
+  HLTOutput.push_back(ch4*gainCorr);// amplitude 
+  HLTOutput.push_back(tsShift4); // time shift of in-time pulse
+  HLTOutput.push_back(ch5); // whatever
+
+}
+
+void HcalDeterministicFit::applyXM(const std::vector<double> & inputCharge, const std::vector<double> & inputPedestal, std::vector<double> & HLTOutput) const {
+
+  std::vector<double> corrCharge;
+  fPedestalSubFxn_.calculate(inputCharge, inputPedestal, corrCharge);
+
+  double TS35[3];
+  double TS46[3];
+  double TS57[3];
+  pulseFraction(inputCharge[3], TS35);
+  pulseFraction(inputCharge[4], TS46);
+  pulseFraction(inputCharge[5], TS57);
+
+  double a3[3] = {TS35[0], TS35[1], TS35[2]};
+  double b3[3] = {0., TS46[0], TS46[1]};
+  double c3[3] = {0., 0., TS57[0]};
+  double d3[3] = {corrCharge[3], corrCharge[4], corrCharge[5]};
+
+  double deno3 = det3(a3, b3, c3);
+
+  double A3 = det3(d3, b3, c3) / deno3;
+  double A4 = det3(a3, d3, c3) / deno3;
+  double A5 = det3(a3, b3, d3) / deno3;
+  
+  HLTOutput.clear();
+  HLTOutput.push_back(A3);
+  HLTOutput.push_back(A4);
+  HLTOutput.push_back(A5);
+
+}
+
+constexpr float HcalDeterministicFit::landauFrac[];
+// Landau function integrated in 1 ns intervals
+//Landau pulse shape from https://indico.cern.ch/event/345283/contribution/3/material/slides/0.pdf
+//Landau turn on by default at left edge of time slice 
+// normalized to 1 on [0,10000]
+void HcalDeterministicFit::getLandauFrac(float tStart, float tEnd, float &sum) const{
+
+  if (abs(tStart-tEnd-25)<0.1) {
```

Comment:
``` std::abs ``` here seems more appropriate, unless the intention was to convert to integer values (in the latter case, convert with an explicit ``` int diffT = tStart - tEnd - 25; diffT = std::abs(diffT); ... ```